### PR TITLE
Add CSVSource model

### PR DESCRIPTION
**Pull request by CFarcy**

### CODE REVIEW

File: django_geosource/models.py [hunk 1]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+    decimal_separator_field = models.CharField(
+        max_length=100, choices=DECIMAL_SEPARATORS, default=POINT[1]
+    )
+    separator_field = models.CharField(
+        max_length=100, choices=SEPARATORS, default=SEMI_COLON[1]
+    )
+    delimiter_field = models.CharField(
+        max_length=100, choices=DELIMITERS, default=QUOTATION_MARK[1]
+    )
+    longitude_field = models.FloatField()
```
Comment (jrmi) [inline django_geosource/models.py:403]: Why it is a floatfield ? It's a field name no ?

File: django_geosource/models.py [hunk 2]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+    SEPARATORS = (COMA, SEMI_COLON, TAB, SPACE, COLUMN)
+    DECIMAL_SEPARATORS = (COMA, SEMI_COLON, POINT, SPACE)
+    DELIMITERS = (QUOTATION_MARK,)
+
+    file = models.FileField(upload_to="geosource/csv/%Y")
+    encoding_field = models.CharField(max_length=100)
+    decimal_separator_field = models.CharField(
+        max_length=100, choices=DECIMAL_SEPARATORS, default=POINT[1]
+    )
+    separator_field = models.CharField(
```
Comment (jrmi) [inline django_geosource/models.py:397]: Naming : `separator`

File: django_geosource/models.py [hunk 3]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+        max_length=100, choices=SEPARATORS, default=SEMI_COLON[1]
+    )
+    delimiter_field = models.CharField(
```
Comment (jrmi) [inline django_geosource/models.py:400]: `delimiter`

File: django_geosource/models.py [hunk 4]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+    QUOTATION_MARK = ("quotationmark", '"')
+    POINT = ("point", ".")
+
+    SEPARATORS = (COMA, SEMI_COLON, TAB, SPACE, COLUMN)
+    DECIMAL_SEPARATORS = (COMA, SEMI_COLON, POINT, SPACE)
+    DELIMITERS = (QUOTATION_MARK,)
+
+    file = models.FileField(upload_to="geosource/csv/%Y")
+    encoding_field = models.CharField(max_length=100)
+    decimal_separator_field = models.CharField(
```
Comment (jrmi) [inline django_geosource/models.py:394]: `decimal_separator`

File: django_geosource/models.py [hunk 5]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+    COLUMN = ("column", " ")
+    QUOTATION_MARK = ("quotationmark", '"')
+    POINT = ("point", ".")
+
+    SEPARATORS = (COMA, SEMI_COLON, TAB, SPACE, COLUMN)
+    DECIMAL_SEPARATORS = (COMA, SEMI_COLON, POINT, SPACE)
+    DELIMITERS = (QUOTATION_MARK,)
+
+    file = models.FileField(upload_to="geosource/csv/%Y")
+    encoding_field = models.CharField(max_length=100)
```
Comment (jrmi) [inline django_geosource/models.py:393]: encoding

File: django_geosource/models.py [hunk 6]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+    )
+    separator_field = models.CharField(
+        max_length=100, choices=SEPARATORS, default=SEMI_COLON[1]
+    )
+    delimiter_field = models.CharField(
+        max_length=100, choices=DELIMITERS, default=QUOTATION_MARK[1]
+    )
+    longitude_field = models.FloatField()
+    latitude_field = models.FloatField()
+    number_lines_to_ignore_field = models.PositiveIntegerField(default=0)
```
Comment (jrmi) [inline django_geosource/models.py:405]: Why do you add field to all field ?

File: django_geosource/models.py [hunk 7]
```diff
@@ -373,3 +374,53 @@ def refresh_data(self):
...
+
+    def get_file_as_sheet(self):
+        with open(self.file.name, "r") as f:
+            try:
+                return pyexcel.get_sheet(
+                    file_type="csv", file_content=f, delimiter=self.separator_field
+                )
+            except pyexcel.exceptions.FileTypeNotSupported:
+                logger.info("Source's CSV file is not valid")
```
Comment (jrmi) [inline django_geosource/models.py:414]: It's just an info log ?

Comment (jrmi) [reply django_geosource/models.py:414]: Does this information have a chance to be read by the end user ?

File: django_geosource/tests/test_models.py [hunk 8]
```diff
@@ -247,3 +249,21 @@ def test_get_status(self):
...
+class ModelCSVSourceTestCase(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.source_name = os.path.join(
+            settings.BASE_DIR, "django_geosource", "tests", "source.csv"
+        )
+        cls.source = CSVSource.objects.create(
+            file=cls.source_name,
+            geom_type=8,
+            longitude_field=42.2222,
```
Comment (jrmi) [inline django_geosource/tests/test_models.py:263]: But why ????

File: django_geosource/models.py (hunk 1 above)
Comment (CFarcy) [reply django_geosource/models.py]: Field has been removed, this data is now stored in the `settings` field.

File: django_geosource/models.py (hunk 2 above)
Comment (CFarcy) [reply django_geosource/models.py]: Field has been removed, this data is now stored in the `settings` field

File: django_geosource/models.py (hunk 3 above)
Comment (CFarcy) [reply django_geosource/models.py]: Field has been removed, this data is now stored in the `settings` field

File: django_geosource/models.py (hunk 4 above)
Comment (CFarcy) [reply django_geosource/models.py]: Field has been removed, this data is now stored in the `settings` field

File: django_geosource/models.py (hunk 5 above)
Comment (CFarcy) [reply django_geosource/models.py]: Field has been removed, this data is now stored in the `settings` field

File: django_geosource/models.py (hunk 6 above)
Comment (CFarcy) [reply django_geosource/models.py]: Fields has been removed.

File: django_geosource/tests/test_models.py (hunk 8 above)
Comment (CFarcy) [reply django_geosource/tests/test_models.py]: Changes/rework have been made to reflect more the functional needs.

File: django_geosource/models.py [hunk 9]
```diff
@@ -373,3 +374,148 @@ def refresh_data(self):
...
+                is_xy = self.settings["coordinates_field_count"] == "xy"
+                lnglat_field = self.settings["latlong_field"]
+
+                # if no header, we except index for the column has been provided
+                if self.settings.get("header"):
+                    coords = row[sheet.colnames.index(lnglat_field)]
+                else:
+                    coords = row[sheet.colnames[int(lnglat_field)]]
+
+                # some fools use a reversed cartesian coordinates system (╯°□°)╯︵ ┻━┻
```
Comment (jrmi) [inline django_geosource/models.py:440]: +1

File: django_geosource/models.py [hunk 10]
```diff
@@ -373,3 +374,148 @@ def refresh_data(self):
...
+                            for i, (name, value) in enumerate(zip(sheet.colnames, row))
+                            if i != row.index(coords)
+                        },
+                    }
+                )
+
+        return records
+
+    def _get_separator(self, name):
+        SEPARATORS = {
```
Comment (jrmi) [inline django_geosource/models.py:458]: separator should be declared at upper level.

File: django_geosource/models.py [hunk 11]
```diff
@@ -373,3 +374,148 @@ def refresh_data(self):
...
+                            if i not in (row.index(x), row.index(y))
+                        },
+                    }
+                )
+            else:
+                sep = self.settings["coordinates_separtor"]
+                is_xy = self.settings["coordinates_field_count"] == "xy"
+                lnglat_field = self.settings["latlong_field"]
+
+                # if no header, we except index for the column has been provided
```
Comment (jrmi) [inline django_geosource/models.py:434]:
```suggestion
                # if no header, we expect index for the column has been provided
```

File: django_geosource/serializers.py [hunk 12]
```diff
@@ -263,3 +267,102 @@ class WMTSSourceSerialize(SourceSerializer):
...
+
+class CSVSourceSerializer(FileSourceSerializer):
+    coordinate_reference_system = CharField(required=True)
+    encoding = CharField(required=True)
+    field_separator = CharField(required=True)
+    decimal_separator = CharField(required=True)
+    char_delimiter = CharField(required=True)
+    coordinates_field = CharField(required=True)
+    number_lines_to_ignore = IntegerField(required=True)
+    header = BooleanField(required=True)
```
Comment (jrmi) [inline django_geosource/serializers.py:280]: naming: use_header ?

File: django_geosource/models.py (hunk 10 above)
Comment (CFarcy) [reply django_geosource/models.py:458]: Do you mean at upper class level or file level?

File: django_geosource/serializers.py (hunk 12 above)
Comment (CFarcy) [reply django_geosource/serializers.py]: Done.

File: django_geosource/models.py [hunk 13]
```diff
@@ -373,3 +374,148 @@ def refresh_data(self):
...
+        sheet = self.get_file_as_sheet()
+        if self.settings.get("header"):
+            sheet.name_columns_by_row(0)
+
+        limit = limit if limit else len(sheet)
+
+        records = []
+        srid = self.settings["coordinate_reference_system"]
+        for row in sheet:
+            if self.settings["coordinates_field"] == "two_columns":
```
[the remark this answers is not in the document]
Comment (CFarcy) [reply django_geosource/models.py:405]: Done.

File: django_geosource/models.py (hunk 10 above)
Comment (CFarcy) [reply django_geosource/models.py]: I put `SEPARATORS` at upper level in class.

File: django_geosource/models.py (hunk 11 above)
Comment (CFarcy) [reply django_geosource/models.py]: I corrected it already :disappointed:

File: django_geosource/models.py (hunk 7 above)
Comment (CFarcy) [reply django_geosource/models.py]: The Exception is re raised with this message now.

**Pull request merged**